[WIP] Use `Diagnostic`s for rendering formatting results

ntBre opened the pull request:
## Summary

This is a rough draft/prototype of using the new `Diagnostic` type for rendering formatter diagnostics. This allows the formatter to inherit all of the output formats already implemented in the linter and ty. For example, here's the new `full` output format, with the formatting diff displayed using the same infrastructure as the linter:

<img width="592" height="364" alt="image" src="https://github.com/user-attachments/assets/6d09817d-3f27-4960-aa8b-41ba47fb4dc0" />

There are several limitiations/todos here still, especially around the `OutputFormat` type:
- A few literal `todo!`s for the remaining `OutputFormat`s without matching `DiagnosticFormat`s
- The default output format is `full` instead of something more concise like the current output
- Some of the output formats (namely JSON) have information that doesn't make much sense for these diagnostics

A solution to all of these could just be using a separate `OutputFormat` type from the one in the linter. I reused that for now just for convenience.

More broadly, it seems pretty expensive to add the entire file's contents as an `Edit`. I'm guessing that might show up in the benchmarks on this PR. On a related note, this is a pretty shallow conversion, only constructing the `Diagnostic`s right before rendering. There might be a better way to use the new infrastructure more. We could also use them for rendering `FormatCommandError`s, not just `FormatPathResult`s.

## Test Plan

Some new CLI tests and manual testing. Another TODO is adding something like the `output_format` test for the linter with snapshots for all of the supported output formats.

MichaReiser commented:
> More broadly, it seems pretty expensive to add the entire file's contents as an Edit. I'm guessing that might show up in the benchmarks on this PR. On a related note, this is a pretty shallow conversion, only constructing the Diagnostics right before rendering. There might be a better way to use the new infrastructure more. We could also use them for rendering FormatCommandErrors, not just FormatPathResults.

A solution here could be to add a `diff` field to `Diagnostic` so that the "edit" is computed lazily or the diff is rendered directly instead of using the edit rendering.


> The default output format is full instead of something more concise like the current output

Hmm, that's an interesting find. This needs some design work.

ntBre commented:
I think this is mostly ready for review now. I did a lot of squashing today to make it easier to review commit-by-commit. 

The first 7 commits are all very small, standalone refactors used in later steps. The 8th commit is the biggest, converting `FormatPathResult`s into diagnostics. The 9th commit is an incremental improvement on that to split notebook `Edit`s by cell so that they can actually be rendered in the `full` output format (otherwise they never satisfy [this check](https://github.com/astral-sh/ruff/blob/main/crates/ruff_db/src/diagnostic/render/full.rs#L141)). Finally, the last commit emits errors as diagnostics too. 

I still have a bunch of TODO comments on the errors. I did my best to match up the error variants with `DiagnosticId`s, including adding two new `DiagnosticId`s, but I think there's still room for improvement. I should probably also add a test for some or all of these.

I haven't tried to truncate the lines in the diff yet either. I know that came up in the design discussion, so I'm happy to tackle it now or leave it for a follow up if desired.

Hopefully it's not too glaring, but in the screenshot in the summary you can see that the `|` for line numbers doesn't line up with the middle of the filename `-->` like it does for lint diagnostics. I don't think there's really a good way around this since the arrow alignment comes from `annotate-snippets` and only gets indented if `annotate-snippets` renders line numbers too. I'm sure we could hack another option into `annotate-snippets`, but it would likely make https://github.com/astral-sh/ruff/issues/20411 harder still.

ntBre commented:
I'm marking this ready for review. I pushed a few more commits resolving the main issues I noted above:
- I added a `lineno_offset`/`header_offset` field for improving the `-->` alignment:
  
  <img width="555" height="417" alt="image" src="https://github.com/user-attachments/assets/829cce57-0fe2-4ad1-b5b7-6f8b77fd2a94" />

  It can still go wrong and makes it harder to un-fork `annotate-snippets`, so I'm happy to revert this, but I think it's a visual improvement in general.
- I added tests for all of the `FormatCommandError` variants and fixed a couple of oversights in their formatting, including factoring out and using some of ty's panic rendering. There are still TODOs on a few of these, but I think I could use some input on the best `DiagnosticId`s to use. The tests at least show what they look like to help us iterate on them.

I think the performance hit is acceptable. It's only ~10 ms (~5%) for the `concise` output on a large project, and the larger discrepancy between `full` output and `--diff` seems justified given the additional work of computing more granular diffs. I am seeing slightly worse performance today compared to the last time I ran the benchmarks, so it may be worth a bit more profiling, but I think the point still stands.

<details><summary>New benchmark table</summary>
<p>

| Command           |     Mean [ms] | Min [ms] | Max [ms] |     Relative |
| :---------------- | ------------: | -------: | -------: | -----------: |
| `stable`          |   200.0 ± 2.3 |    196.6 |    203.5 |         1.00 |
| `preview-full`    | 9134.8 ± 32.1 |   9108.0 |   9197.5 | 45.67 ± 0.55 |
| `preview-concise` |   232.6 ± 4.2 |    227.0 |    240.7 |  1.16 ± 0.02 |
| `stable-diff`     | 3354.8 ± 15.9 |   3333.7 |   3383.6 | 16.77 ± 0.21 |

`stable` is basically identical to last time, but both `preview` versions are ~20 ms slower. `--diff` is almost 50 ms slower, which really doesn't make sense.

</p>
</details> 

Similarly, I think we may want to truncate large diffs at some point (Zanie even mentioned possibly making the limit configurable), but I think we can hold off on that for now.

MichaReiser commented on crates/ruff_db/src/diagnostic/mod.rs:
```diff
@@ -1016,6 +1032,17 @@ pub enum DiagnosticId {
+    /// The code needs to be formatted.
+    Unformatted,
+
+    /// Range formatting isn't supported for notebooks.
+    RangeFormatNotebook,
```
I don't think it's worth having a diagnostic ID for every CLI option that can possibly be invalid. Maybe `invalid-cli-option`?

MichaReiser commented on crates/ruff_db/src/diagnostic/mod.rs:
```diff
@@ -1016,6 +1032,17 @@ pub enum DiagnosticId {
+    /// Range formatting isn't supported for notebooks.
+    RangeFormatNotebook,
+
+    /// Something went wrong while formatting.
+    ///
+    /// This often indicates a bug in the formatter.
+    FormatError,
```
Similar to `RangeFormatNotebook`. I wonder if we should just have one `internal-error` code. The main advantage I see is that it makes it very clear that this is a bug in Ruff vs. some user error (we should probalby add a note saying that too)

MichaReiser reviewed (commented):
Nice: I've a few suggestions. The part I feel most unsure about is if using `set_file_level` is correct for format diagnostics.

ntBre commented on crates/ruff/src/commands/format.rs:
```diff
@@ -566,14 +611,106 @@ impl<'a> FormatResults<'a> {
 
+    /// Write a list of the files that would be changed and any errors to the given writer.
+    ///
+    /// Errors are reported first in the order they are provided, followed by the remaining
+    /// diagnostics sorted by file name.
```
I was just trying to preserve the stable behavior of emitting errors first, but it makes sense to me to sort them all together here.

ntBre commented on crates/ruff/src/commands/format.rs:
```diff
@@ -566,14 +611,106 @@ impl<'a> FormatResults<'a> {
+
+        let context = EmitterContext::new(&notebook_index);
+        let config = DisplayDiagnosticConfig::default();
+        match output_format {
+            OutputFormat::Concise => {
+                let config = config
+                    .format(DiagnosticFormat::Concise)
+                    .hide_severity(true)
+                    .color(!cfg!(test) && colored::control::SHOULD_COLORIZE.should_colorize());
+                let value = DisplayDiagnostics::new(&context, &config, &diagnostics);
+                write!(f, "{value}")?;
+            }
+            OutputFormat::Full => {
+                let config = config
+                    .format(DiagnosticFormat::Full)
+                    .hide_severity(true)
+                    .show_fix_diff(true)
+                    .color(!cfg!(test) && colored::control::SHOULD_COLORIZE.should_colorize());
+                let value = DisplayDiagnostics::new(&context, &config, &diagnostics);
+                write!(f, "{value}")?;
+            }
```
The snippet in your comment is the main difference. In the linter we still use the `TextEmitter` for these output formats and mix in a few other checks in the `match`:

https://github.com/astral-sh/ruff/blob/2af8c53110227598cb4758a777667dbe18bc7f15/crates/ruff/src/printer.rs#L251-L268

We also use at least one different configuration setting in the formatter to always suppress the fix availability icon.

ntBre commented on crates/ruff/src/commands/format.rs:
```diff
@@ -664,6 +889,91 @@ impl FormatCommandError {
+                let mut diagnostic = Diagnostic::new(
+                    DiagnosticId::Panic,
+                    Severity::Fatal,
+                    panic_error.to_diagnostic_message(path.as_ref().map(|path| path.display())),
```
Yes, I think so. I liked how ty rendered panics, so I pulled from there instead, but I can either switch to how the linter constructs panics or port this to the linter too.

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -628,6 +765,94 @@ impl<'a> FormatResults<'a> {
+                let source_file = SourceFileBuilder::new(path, unformatted.source_code()).finish();
+                let span = Span::from(source_file);
+                let mut annotation = Annotation::primary(span);
+                annotation.set_file_level(true);
```
Oh, I see. The difference is that we don't actually render a code snippet, we only render a fix (which is something different in annotated snippet?)

And the issue is that the rendering of `-->` isn't aligned with the rendering of the fix.

ntBre replied:
Yep, the fixes/diffs aren't actually rendered through `annotate-snippets`, they're my best attempt to emulate `annotate-snippets` after we render the rest of the diagnostic. That's why the arrow header can't align itself, the two steps are separate.

ntBre replied:
> I see. I think we should rename `set_file_level` to `hide_snippet` or similar to make the distinction clear now that it's used for more than just hiding snippets for file level diagnostics.

I guess I was still thinking of these conceptually as file-level diagnostics. That's what they are on stable anyway, "File would be reformatted." But I'm happy to rename the setting, `hide_snippet` makes the actual effect more clear for sure.

> For the alignment. We can probably get it "right" if we compute the line number where the first change is.

That's roughly what I'm doing now, except it's a bit better to compute the line number of the _last_ change (currently the end of the file) since that's likely to need a wider alignment than the start. Notebooks are also more complicated since the width is for the per-cell line number.

> An alternative would be to remove the need for the alignment altogether. Unfortunately, this might be another breaking change except if all we do is change the number of whitespace before after the arrow

That's interesting. I do think it looks nice aligning the middle of the arrow with the line number separator, but maybe it's not worth the trouble.

ntBre replied:
I renamed the file-level references to `hide_snippet` and updated my todo about removing it. I also added an expanded new todo about the offset hack for now.

MichaReiser commented on crates/ruff/src/args.rs:
```diff
@@ -537,6 +537,11 @@ pub struct FormatCommand {
     #[arg(long, help_heading = "Miscellaneous", alias = "exit-non-zero-on-fix")]
     pub exit_non_zero_on_format: bool,
+
+    /// Output serialization format for violations.
+    /// The default serialization format is "full".
```
Yeah, warning on the option seems difficult.

ntBre commented on crates/ruff/src/commands/format.rs:
```diff
@@ -628,6 +765,94 @@ impl<'a> FormatResults<'a> {
+                );
+
+                let path = result.path.to_string_lossy();
+                // For now, report the edit as a replacement of the whole file's contents. For
+                // scripts, this is a single `Edit`, but notebook edits must be split by cell in
+                // order to render them as diffs.
```
Sounds good, I had an earlier version that did this before I rebased over it. I'll add it back.

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -628,6 +739,127 @@ impl<'a> FormatResults<'a> {
+        notebook_index: &mut FxHashMap<String, NotebookIndex>,
+    ) -> impl Iterator<Item = Diagnostic> {
+        self.results.iter().filter_map(|result| {
+            let FormatResult::Diff {
+                unformatted,
+                formatted,
+            } = &result.result
+            else {
+                return None;
+            };
+
```
Should we use a match here that explicitly skips over `Unchanged` and `Skipped`, panics for `Formatted` (at least in debug builds), and returns `unformatted` and `formatted` for diff?

ntBre replied:
Sure, I was just following the logic in `write_changed`, but a debug assertion makes sense to me.

https://github.com/astral-sh/ruff/blob/00c8851ef8d643434340577f1b1ee65b5e728ac1/crates/ruff/src/commands/format.rs#L555-L559

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -628,6 +739,127 @@ impl<'a> FormatResults<'a> {
+                (fix, line_count)
+            };
+
+            // Locate the first and last characters that differ to use as the diagnostic
+            // range.
+            let range = {
+                let mut start = None;
+                let mut end = None;
+                for ((offset, old), new) in unformatted
+                    .source_code()
+                    .char_indices()
+                    .zip(formatted.source_code().chars())
+                {
+                    if old != new {
+                        if start.is_none() {
+                            start = Some(offset);
+                        } else {
+                            end = Some(offset);
+                        }
+                    }
+                }
```
I think we should use this range for the Edit as well if we already computed it anyway. It can help to reduce the size of the edit in common cases where only a few files are incorrectly formatted

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -628,6 +739,127 @@ impl<'a> FormatResults<'a> {
+
+            // Locate the first and last characters that differ to use as the diagnostic
+            // range.
+            let range = {
```
I think we can do better than traversing the entire file by:

1. Search from the beginning and find the first changed character
2. Slice the remaining document to `&source[start..]`
3. Search from the end to find the last changed character


This is still `O(n)` in the worst case, but should do much better if a large portion of the file changed. It might be worth extracting this logic into a helper function to not distract the reading flow

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -628,6 +739,127 @@ impl<'a> FormatResults<'a> {
+            // `ruff_db::diagnostic::render::full` into `annotate-snippets`, likely as another
+            // `DisplayLine` variant and update the `lineno_width` calculation in
+            // `DisplayList::fmt`. That would handle this offset "automatically."
+            let lines = OneIndexed::new(line_count as usize).unwrap_or_default();
```
We should use the line number of the `Edit`'s end + rendered context window size. The arrow is otherwise over intended if only the first line changed in a 1000 lines file (which I think looks worse than always aligning with two spaces).

ntBre replied:
Ah good call. This was a bit surprising to me, so I just wanted to note it here: I think the context window size is up to 6 lines. We pass `3` to similar's `grouped_ops` function here:

https://github.com/astral-sh/ruff/blob/1d3e4a91534448121c6db99c2baf83c1ccd2d909/crates/ruff_db/src/diagnostic/render/full.rs#L157-L159

which is used in similar [here](https://github.com/mitsuhiko/similar/blob/a169e29954502d67e26c9e766f2f719e99a52559/src/common.rs#L134-L136):

```rust
            // End the current group and start a new one whenever
            // there is a large range with no changes.
            if len > n * 2 {
```

I guess in a strict sense we should also `min` this with the actual number of lines in the file, so we don't end up with 105 in a 99 line file, for example.

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -658,12 +889,65 @@ impl FormatCommandError {
+                Diagnostic::new(DiagnosticId::Io, Severity::Error, source_error)
+            }
+            FormatCommandError::Format(_, format_module_error) => match format_module_error {
+                FormatModuleError::ParseError(parse_error) => Diagnostic::new(
+                    DiagnosticId::InvalidSyntax,
+                    Severity::Error,
+                    &parse_error.error,
+                ),
+                FormatModuleError::FormatError(format_error) => {
```
Nit: I'd also classify this as an internal error. At least, how we're using it in the formatter. 

The way we use this error variant is to avoid panicking if we didn't see a specific syntax element when doing some ad-hoc lexing. This means, we incorrectly assumed some invariant that isn't true because the formatter is never called on code with syntax errors. This is always an implementation error -> `InternalError`)

The original intended use case for this error variant comes from biome where the formatter supports formatting code with syntax error. The way it works is that it formats all code without syntax errors. For sections that contain syntax errors, it falls back to preserving the formatting from the source by catching `FormatError::SyntaxError` at the statement level. But even for Biome, a `SyntaxError` should never be propagated passt the statement formatting. If it does, than that's consider an internal error too.

MichaReiser reviewed (approved):
I've a few more suggestions about the computed diagnostic and `Edit` range and on how to compute the offset for the arrow.

I still very much dislike the offset hack but I think we can get it "correct" enough outside annotated snippet. However, it might be worth spending like a day to see if we can move the diff rendering to annotated snippets, removign the need for the offset workaround.

ntBre commented:
Thank you for the reviews! 

I think this could use one more look. I was over-complicating the range calculations for a while today, but they felt kind of tricky, at least when I was trying to use `zip` like the `start` calculation[^1]. The new `ModifiedRange` type seems to be working well now, though, unless I missed any edge cases.

The other changes seemed relatively straightforward, and I also merged the changes from #20595.

I'll timebox trying to move our diff rendering to annotate-snippets to one day later this week.

[^1]: The tricky part was that zipping and `find`ing the first different character could easily fail if one of the snippets was shorter than the other; at that point it's not clear which one was shorter and caused the failure. We also want an exclusive range, so we've gone one character too far by finding the first different character rather than the last common character. Now I just loop from the end and track the length of the common suffix, which we can subtract from both `text_len`s despite the actual offsets likely being different.

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -628,6 +670,155 @@ impl<'a> FormatResults<'a> {
+                (fix, line_count)
+            } else {
+                let formatted_code = &formatted.source_code()[modified_range.formatted];
+                let edit = if formatted_code.is_empty() {
```
It shouldn't matter so I think it's fine leaving as is but there's the third case where `modified_range.unformatted` is empty (e.g. when adding blank lines between two classes), in which case its an insertion. We could add a `Edit::from_text_and_range(new_text, range)` (with a better name) that does this dance.

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -628,6 +670,155 @@ impl<'a> FormatResults<'a> {
+            // `ruff_db::diagnostic::render::full` into `annotate-snippets`, likely as another
+            // `DisplayLine` variant and update the `lineno_width` calculation in
+            // `DisplayList::fmt`. That would handle this offset "automatically."
+            let line_count = (line_count + CONTEXT_LINES).min(
+                formatted
+                    .source_code()
+                    .count_lines(TextRange::up_to(formatted.source_code().text_len())),
```
Is using `formatted` here correct? What if unformatted is shorter?

ntBre replied:
We discussed this in our 1:1, but we only display line numbers for the new code, so I think this is correct. Somewhat random lint rule example:

```
help: Replace for loop with list comprehension
1 | original = list(range(10000))
  - filtered = []
  - for i in original:
  -     if i % 2:
  -         filtered.append(i)
2 + filtered = [i for i in original if i % 2]
```

This is a deletion but still shows that we only render  the new line numbers.

MichaReiser commented on crates/ruff/src/commands/format.rs:
```diff
@@ -792,6 +1018,56 @@ impl Display for FormatCommandError {
+        let unformatted = unformatted.source_code();
+        let formatted = formatted.source_code();
+
+        let start = unformatted
+            .char_indices()
+            .zip(formatted.chars())
+            .find_map(|((offset, old), new)| {
+                (old != new).then_some(TextSize::try_from(offset).unwrap())
+            })
+            // Fall back on the shorter text length if one of the strings is a strict prefix of the
+            // other (i.e. the zip iterator ended before finding a difference).
+            .unwrap_or_else(|| unformatted.text_len().min(formatted.text_len()));
```
I wonder if a regular loop would have been easier here (similar to what you have below):

```rust
let mut prefix_length = TextSize::ZERO;

for (unformatted, formatted) in unformatted.chars().zip(formatted.chars()) {
    if unformatted != formatted {
        break;
    }

    prefix_length += unformatted.text_len();
}
```

tallerasaf commented:
@ntBre 

Now, `ruff format --check`, become like `ruff format --diff`? But with more features.

So, `ruff format --diff` is deprecated for preview mode now?

And why the docs don't mention it?

```
      --check
          Avoid writing any formatted files back; instead, exit with a non-zero
          status code if any files would have been modified, and zero otherwise
      --diff
          Avoid writing any formatted files back; instead, exit with a non-zero
          status code and the difference between the current file and how the
          formatted file would look like
```

Looks like the `--check` CLI flag didn't change in the docs, But it did.

ntBre commented:
`--diff` produces a standalone diff that can still be useful for applying as a patch, for example, so it's not deprecated. The "diff" shown by the default `format --check` output uses the same format that's in preview for lint rules, which is a bit different.

I also think the `--check` help message is still accurate. The new information is in the new `--output-format` entry in the CLI help.
